src/neoxp/Node/NodeUtility.cs

```diff
@@ -208,6 +211,186 @@ public static void SignOracleResponseTransaction(ProtocolSettings settings, Expr
+                try
+                {
+                    stateHeight_ = await stateAPI.GetStateHeightAsync();
+                }
+                catch (RpcException e)
+                {
+                    if (e.Message.Contains("Method not found"))
+                    {
+                        throw new Exception(
+                            "Could not get state information. Make sure the remote RPC server has state service support");
+                    }
+                    throw;
+                }
+                
+                if (stateHeight_.localRootIndex is null)
+                {
+                    throw new Exception("Null \"localRootIndex\" in state height response");
+                }
+                height = stateHeight_.localRootIndex.Value;
+            }
+
+            var stateRoot = await stateAPI.GetStateRootAsync(height);
+
+            // rpcClient.GetContractStateAsync returns the current ContractState, but this method needs
+            // the ContractState as it was at stateHeight. ContractManagement stores ContractState by
+            // contractHash with the prefix 8. The following code uses stateAPI.GetStateAsync to retrieve
+            // the value with that key at the height state root and then deserializes it into a ContractState
+            // instance via GetInteroperable.
+
+            var key = new byte[21];
+            key[0] = 8; // ContractManagement.Prefix_Contract
+            _contractHash.ToArray().CopyTo(key, 1);
+
+            var contractStateBuffer = await stateAPI.GetStateAsync(
+                stateRoot.RootHash, NativeContract.ContractManagement.Hash, key);
+            var contractState = new StorageItem(contractStateBuffer).GetInteroperable<ContractState>();
+            var states = await rpcClient.ExpressFindStatesAsync(stateRoot.RootHash, _contractHash, new byte[0]);
+
+            return (contractState, states);
+        }
+
+        public static int PersistContract(SnapshotCache snapshot, ContractState state,
+            IReadOnlyList<(string key, string value)> storagePairs, ContractCommand.OverwriteForce force)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_Contract).Add(state.Hash);
+            var localContract = snapshot.GetAndChange(key)?.GetInteroperable<ContractState>();
+
+            if (localContract is null)
+            {
+                // Our local chain might already be using the contract id of the pulled contract, we need to check for this
+                // to avoid having contracts with duplicate id's. This is important because the contract id is part of the
+                // StorageContext used with Storage syscalls and else we'll potentially override storage keys or iterate
+                // over keys that shouldn't exist for one of the contracts.
+                if (state.Id <= LastUsedContractId(snapshot))
```

Comment:
Nit: Why not just always use the next available ID? Since we have to handle the case where the local id != remote id, why not simplify the logic?

Reply:
I tried to keep it the same as the remote contract whenever possible. Always using the last is also an option.

Reply:
I think it's simpler just to issue a new local contract id